src/seqn/trans.lisp

```diff
@@ -0,0 +1,399 @@
+(in-package :geb.seqn.trans)
+
+(defun remove0-dom (morphism)
+  (remove 0 (dom morphism)))
+
+;; Add range-n function
+
+(defun range-constraints-dom (domain)
+  (cond ((zerop (length domain))   nil)
+        ((= (length domain) 1)    (list (geb.vampir:range-n
+                                         (vamp:make-constant :const (car domain))
+                                         (vamp:make-wire :var
+                                                         (intern (format nil "x~a" 1))))))
+        (t (cons (geb.vampir:range-n (vamp:make-constant :const (car (last domain)))
+                                (vamp:make-wire :var
+                                                (intern (format nil "x~a" (length domain)))))
+            (range-constraints-dom (butlast
+                                    domain))))))
+
+(defmethod to-circuit ((morphism <seqn>) name)
+  "Turns a SeqN term into a Vamp-IR Gate with the given name
+Note that what is happening is that we look at the domain of the morphism
+and skip 0es, making non-zero entries into wires"
+  (let* ((wire-count (length (dom morphism)))
+         (wires (loop for i from 1 to wire-count
+                      collect (vamp:make-wire :var (intern (format nil "x~a" i)
+                                                           :keyword)))))
+    (list
+     (vamp:make-alias :name name
+                      :inputs wires
+                      :body  (append (range-constraints-dom (dom morphism))
+                                     (list
+                                      (vamp:make-tuples :wires
+                                                        (remove nil (mapcar (lambda (x) (if (= (car x) 0)
+                                                                                            nil
+                                                                                            (cadr x)))
+                                                                            (prod-list (cod morphism)
+                                                                                       (to-vampir morphism
+                                                                                                  wires
+                                                                                                  nil))))
+                                                        )))))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; SeqN to Vamp-IR Compilation
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```

Comment:
General comment on Vampir @agureev :

Maybe we could remove the constant checks into an optimizable check, and have that logic in it's own check.

We could maybe even run it as a pass before this code.

As this code mixes generation with some optimziation, making both read messy. This can be done as a pass or by just doing a check and dispatching so at least it's in a separate place.

Reply:
This is a good point and I agree that mixing generation and optimization is messy, but I think this is something for another patch. I'll leave this untreated for now and address it after we release v.0.5.0 since I am not sure of the most efficient way here. Reducing everything as things generate makes it easier for internal printing to handle stuff.